managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -2230,9 +2236,8 @@ void updateCursor(ManagedCursorImpl cursor, PositionImpl newPosition) {
     PositionImpl startReadOperationOnLedger(PositionImpl position, OpReadEntry opReadEntry) {
         Long ledgerId = ledgers.ceilingKey(position.getLedgerId());
         if (null == ledgerId) {
-            opReadEntry.readEntriesFailed(new ManagedLedgerException.NoMoreEntriesToReadException("The ceilingKey(K key"
-                    + ") method is used to return the least key greater than or equal to the given key, "
-                    + "or null if there is no such key"), null);
+            opReadEntry.makeInvalid();
```

Comment:
This changes the behavior of the method. So we need to go through all the usage of this method.
As far as I can see, 
1. This line makes no sense any more. https://github.com/apache/pulsar/blob/a242f03f69791fcfa1934b6cab20689393228381/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpReadEntry.java#L142
2. Also affects the `OpReadEntry.create` in https://github.com/apache/pulsar/blob/5cf3fa0d8050b16305ee060820b884d69ec8a828/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L765

Reply:
Thanks for your reminder. I have checked it, found another problem, I will fix it at another pr.
This one patch just for npe.

Reply:
location 1:
It's useless code, I remove it at #15104. So needn't check it.

location 2:
It's a wait opReadEntry, when opAddEntry completed, it will nofity wait opReadEntry read again. The logicment is not conflict with this changes. And I fix the bad behavior when wait opReadEntry read at #15104

Reply:
> It's a wait opReadEntry, when opAddEntry completed, it will nofity wait opReadEntry read again.

Can you add a unit test to cover this path?
It seems `readPosition` of this invalid OpReadEntry is used before `asyncReadEntries` in `hasMoreEntries`

Reply:
> Can you add a unit test to cover this path? It seems `readPosition` of this invalid OpReadEntry is used before `asyncReadEntries` in `hasMoreEntries`

Maybe not, it will still be used in asyncReadEntries.

At line_768, the wait OpReadEntry set to property `WAITING_READ_OP_UPDATER`.
https://github.com/apache/pulsar/blob/c1631588fd71f4e5d63557741bacf2b56f27563d/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L768

In notifyEntriesAvailable, get wait OpReadEntry from `WAITING_READ_OP_UPDATER`, and handle it to `asyncReadEntries `. 

And in notifyEntriesAvailable, there be a bad behavior. fixes at #15102

https://github.com/apache/pulsar/blob/c1631588fd71f4e5d63557741bacf2b56f27563d/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java#L2749-L2772

Reply:
Add unit to cover wait opReadEntry case. And close #15102, the handle push at this pr.

Reply:
all done, pls check it again when you convenient.